New major version without Xerces

kuniss opened the pull request:
This PR is intended to implement the new major version 4 of javapos-config-loader incorporating the following major changes
- replacing the Xerces XML parser implementation by a Javax based one (contribution by @mjpcger )
- a lot of source code improvements 
  - generic types, code smell removals, 
  - modernization of the 1999 code base, 
  - internal data structure hiding
- removal of deprecated methods and classes
- solving long standing issue #1 to enable javapos-config-loader for web-start environments

mjpcger committed:
Assignments of typed values via deprecated constructors replaced by
type-casted values.

mjpcger committed:
Initial Javax based XML populator implementation to replace Xerces

The new class JavaxRegPopulator extends the config loader with an XML
populator that works with Javax instead of Xerces and thus enables
JavaPOS to be used without the external modules of the Xalan / Xerces
project.

mjpcger committed:
Replaced Xerces populator by new Javax populator.

The default SimpleXmlRegPopulator has been changed to use
JavaxRegPopulator instead of XercesRegPoputator now. This allows to
remove Xerces completely later on.

mjpcger committed:
Avoid IllegalArgumentExceptions if an optional URL related XML attribute
is missing.

Xerces2RegPopularor did not work correctly whenever one of the optional
attributes are missing. Using an empty string in that case.

mjpcger committed:
Changes made after PR being reviewed

according to discussion in
https://github.com/JavaPOSWorkingGroup/javapos-config-loader/pull/7.

In addition to requested changes, string constant "jpos.xml" has been
replaced by constant DEFAULT_XML_FILE_NAME in method save.

mjpcger committed:
Refactored out string as constants to XmlRegPopulator

- Constants for path and names of DTD and XSD file and for the DTD
document type value moved to XmlRegPopulator.
- Constants for XML tag and attribute values added to XmlRegPopulator.
- "Unchecked" warning suppressed for some methods.
- Method getSortedList removed. Use Collections.sort directly instead.
- Method notAttribute renamed to cannotSetAttributeOfTag for better
understanding.
- Comments for method cannotSetAttributeOfTag added.
- LinkedList replaced by ArrayList where meaningful and vice versa.

kuniss committed:
Using scalar types' .valueOf method instead of auto-boxing

for better readability.

kuniss committed:
Code clean-up: Get rid of raw types and Vector usage at JposEntryUtility

where possible as we are going to change it anyway. However, avoided
changes at public method signature to ensure backward compatibility.

kuniss committed:
Code clean-up: simplified negated boolean expressions

kuniss committed:
Avoiding file resources remaining unclosed for the SimpleRegPopulator

by implementing all resources access using the try-with-resources code
schema where possible and meaningful.

kuniss committed:
Fixing a NPE in case multi-prop definition is missing in jpos.properties

if implicitly requested by usage of the DefaultCompositeRegPopulator.
Instead throwing a runtime exception .

Restricted the check from null pointer check to empty list check,
because in case of an empty list the subsequent code may cause the NPE
seen in the tests after having applied the try-with-resources code
schema strictly.

kuniss committed:
Code cleanup: Got rid of compile warning, introduce generics

where possible. If not possible SuppressWarnig annotation had been
added, e.g. on public APIs.

Furthermore:
- clean up not done for Xerces code as it will be removed
- low hanging fruits taken, e.g. using StringBuilder instead of
StringBuffer
- improved initialization of singletons in jpos.profile
- further code improvements to get less code smell (no complains for
this old code...)

Tests passed locally.

kuniss committed:
Code cleanup: get rid of a dead code warning

by restructuring the control flow.

kuniss committed:
Code cleanup: get rid of some code smell

like
- bad constant referencing
- StringBuffer usage
- unnecessary casting

kuniss committed:
Removed deprecated constructors and deprecated Tracer class

This is a backward incompatible change!
But OK for the new major number 4.0.

All deprecated since long time (1999)
- jpos.config.simple.SimpleEntryRegistry parameterless constructor
- jpos.loader.simple.SimpleServiceManager parameterless constructor
- jpos.util.Tracer internal class

mjpcger committed:
Code cleanup: get rid of some warnings

kuniss committed:
Corrected an initialization error when reloading to the same populator

instance a different file.
Solved by clearing the internal data structure, which was missing.
Detected by adding a copy of XercesRegPopulatorTestCase which
instantiates JavaxRegPopulator and fails for test
jpos.config.simple.xml.JavaxRegPopulatorTestCase.testXercesPopulator1().

kuniss committed:
Code cleanup for the new JavaxRegPopulator test case.

kuniss committed:
Added assertions checking getLastLoadException is null.

Some other source formatting cleanups.

kuniss committed:
Re-implemented method insertJposPropertiesInElement

for a more readable implementation as discussed at
https://github.com/JavaPOSWorkingGroup/javapos-config-loader/pull/7#discussion_r1528472649

kuniss committed:
Code cleanup: Get rid of all javadoc warnigs and errors.

kuniss committed:
Code cleanup: Replaced iterator loop by for-each loop.

Note, there is still a compiler warning at this changed lines requesting
to use generic types. This is already fixed with branch fixes-from-pr7
and will work out when merging both branches together on the main.

kuniss committed:
Restricted access to DTD files while loading JavaPOS XML configuration.

Code cleanup: Defined a constant for that restriction rule attribute.

kuniss committed:
Avoiding XSD resource remains unclosed.

kuniss committed:
Rolled back 2 generic type improvements to ensure backward compatibility

Affected methods are
- findFileInJarZipFiles(String fileName, Vector<String>
jarZipFilesVector) restored, forwarding to the new method
findFileInJarZipFiles( String fileName, List<String> jarZipFilesList )
- getJposEntries() conceptually substituted by methods
addJposEntry(String logicalName, JposEntry entry) and
clearAllJposEntries()

Both old reintroduced methods has been marked as deprecated, however,
pointing to the new methods.

kuniss committed:
Code cleanup: some code formatting improvements.

No semantic changes at all.

kuniss committed:
Get rid of Xerces in the ProfileFactory implementation.

DefaultProfileFactory.java is a copy of XercesProfilefactory from the
fixes-from-pr7 branch coming with a lot of code clean up.

At SimpleServiceManager.java the instantiation of the ProfileFactory has
been changed to the new DefaultProfileFactory.

kuniss committed:
Added test case for DefaultProfileFactory.

Follow up to the previous commit.

kuniss committed:
Finally, removed Xerces dependencies.

kuniss committed:
Code cleanup: Removed all "extends Object" declarations

kuniss committed:
Added change notes related to code improvements

done at branch fixes-from-pr7.

kuniss committed:
Added change notes related to Xerces removal

done at branch xerces-replacement-from-pr7.

kuniss committed:
Merge xerces-replacement-from-pr7 into 4.0-improvements+without-xerces

as preparation for an PR which gets merged into main containing both
efforts.

kuniss committed:
Deleted the conflicting Xerces2RegPopulator.java subsequently

which was not staged during previous merge due to an Eclipse Git tooling
problem.

kuniss committed:
Code cleanup: get rid of deprecated method calls

which got marked as deprecated and replaced by new type-safe and
internal-data-strcuture-protecting calls.
That way the last compiler warning in the merged code got resolved.

kuniss committed:
Enabling for webstart environments by solving issue #1

by using standard class-loader instead of system class loader.

kuniss committed:
Prepares versions for 4.0 releasing.

kuniss commented:
@mjpcger, do you want to review this pull request before I'm merging it? Most of the work is from your contribution...

I would, however, publish a pre-release of the original branch to Maven Central such that it can be verified in local build environments.

mjpcger commented on src/main/java/jpos/config/JposConfigException.java:
```diff
@@ -28,6 +28,11 @@
 public class JposConfigException extends JposException
 {
+	/**
+	 * 
```
JavaDoc comment missing

kuniss replied:
I removed the empty comment. It had been added automatically by Eclipse when generating the serial UID to solve a compiler warning.

mjpcger commented on src/main/java/jpos/config/JposEntryRegistryEvent.java:
```diff
@@ -28,6 +28,11 @@
 public class JposEntryRegistryEvent extends EventObject
 {
     /**
+	 * 
```
Meaningful JavaDoc comment missing

kuniss replied:
I removed the empty comment. It had been added automatically by Eclipse when generating the serial UID to solve a compiler warning.

mjpcger commented on src/main/java/jpos/loader/JposLoaderException.java:
```diff
@@ -32,6 +32,11 @@ public class JposLoaderException extends JposException
 	// 
 	
+	/**
+	 * 
```
Meaningful JavaDoc comment missing

kuniss replied:
I removed the empty comment. It had been added automatically by Eclipse when generating the serial UID to solve a compiler warning.

mjpcger commented on src/main/java/jpos/profile/ProfileException.java:
```diff
@@ -29,18 +29,28 @@ public class ProfileException extends Exception
 	//
 
+	/**
+	 * 
```
Meaningful JavaDoc comment missing

kuniss replied:
I removed the empty comment. It had been added automatically by Eclipse when generating the serial UID to solve a compiler warning.

mjpcger reviewed (commented):
Se.ems to be good, only some JavaDoc comments should not remain empty

kuniss committed:
Code cleanup: I removed empty comments for serial UIDs

as requested by reviewer.
They had been added automatically by Eclipse when generating the serial
UID to solve compiler warnings. A comment would be meaningsless at the
end here...

kuniss commented:
@mjpcger could you please make a final review and approve the PR, please?
You should have something like the following in the GitHub GUI under the _Review Changes_ button:
![image](https://github.com/JavaPOSWorkingGroup/javapos-config-loader/assets/1107971/c50acc4e-e8be-457c-8107-c51182db7c0c)

I could merge without your approval (because I'm the project admin 😄), however, I would like to try the "clean" process way...
Thanks!

kuniss commented:
A RC built from this PR has been published to Maven Central's snapshot repository at https://oss.sonatype.org/content/repositories/snapshots/org/javapos/javapos-config-loader/4.0.0-SNAPSHOT/. 

It may be downloaded from there or resolved by dependency resolution if configured appropriately, e.g. for Gradle:
```
repositories {
    maven {
        url 'https://oss.sonatype.org/content/repositories/snapshots'
    }
}

dependencies {
    implementation 'org.javapos:javapos-config-loader:4.0.0+'
}
```

Who ever is interested, please integrate in your environment for testing purposes and report back any issue here in this PR.

mjpcger reviewed (approved):
Looks good for me, no change necessary.

mjpcger commented:
More approval is not possible for me. Looks as if I have no write access and therefore my approval is not enough to change the **Review required** status.

kuniss commented:
Thanks, @mjpcger!
I adjusted the  branch rules. Now I can merge according to the rules. 😄

kuniss commented:
Who ever is interested: We just encountered in our off-side integration test that some of our JavaPOS XML configuration files are not read in correctly by the new javapos-config-loader-4.0.0-RC1, Those were read in without errors by the Xerces implementation, however.

We are investigating...

kuniss committed:
Replaced legacy trace enabling at tests by new one.

kuniss committed:
Enabled tracing at tests by default.

kuniss committed:
Revert "Enabled tracing at tests by default."

This reverts commit 46b614882d9b281d7dcb53f2b02f2979b62a5507.

kuniss committed:
Enabled tracing on tests by default.

kuniss committed:
Improved tracing to log the valuable information.

kuniss commented:
The reported regression was a misinterpretation. In fact, it new XML populator is just more accurate. If the DTD is given in the JavaPOS XML configuration file like 
´´´
<!DOCTYPE JposEntries PUBLIC "-//JavaPOS//DTD//EN"
                             "jpos/res/jcl.dtd">
´´´
then file _jpos/res/jcl.dtd_ must be in the classpath, otherwise the validation fails and no entries are read in.

kuniss commented:
I improved the tracing a bit. So, I'm going to publish a new SNAPSHOT on Maven Central's snapshot repository - RC2.

kuniss committed:
Trim tracer names from jpos.util.tracing.TurnOnNamedTracers definition.

kuniss committed:
Code clean up for some generic type issues.

kuniss committed:
Replaced XercesProfileFactory.java by DefaultProfileFactory.java

which is the same implementation, just another name to get rid of
"Xerces" in the name.
In fact, it was forgotten to remove the old file when renaming it.

kuniss committed:
load() at composite populators throw a more specific runtime exception

in case there is a configuration error which can not be recovered from.

kuniss committed:
Code cleanup: made code less complex

by refactoring out code to private methods for better readability.

kuniss committed:
Code improvements regarding try-with-resouce and Files.delete() in
XMLHelper.java
